File: src/public/javascripts/menu.js
/**
 Copyright 2011 Red Hat, Inc.

 This software is licensed to you under the GNU General Public
 License as published by the Free Software Foundation; either version
 2 of the License (GPLv2) or (at your option) any later version.
 There is NO WARRANTY for this software, express or implied,
 including the implied warranties of MERCHANTABILITY,
 NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
 have received a copy of GPLv2 along with this software; if not, see
 http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
 */

/*
 * Katello JavaScript Menu File
 * Author: jrist
 * Date: 10/25/11
 * Time: 10:19 AM
 */

/**
 * Document Ready function
 */
$(document).ready(function (){
  KT.menu.menuSetup();
  KT.menu.thirdLevelNavSetup();
});

KT.menu = (function(){
  return {
    menuSetup: function(){
      //set some useful vars
      var topLevel = $('nav.tab_nav li.top_level');
      var activeSubnav = $('nav.tab_nav .second_level').filter(".selected").parent();
      var secondLevel = $('nav.tab_nav .second_level').parent();
      var subnav = $('#subnav');

      var activeTab = topLevel.filter('.selected:visible');

      //hide the secondlevel before prepending it to subnav
      secondLevel.hide();
      //set the current tab to active so we can check it later
      activeTab.addClass('active');
      activeTab = topLevel.filter('.active');
      //move (prepend) the second level nav items to subnav
      secondLevel.prependTo(subnav);

      //some settings for the hoverable top level tabs
      var hoverSettings = {
        sensitivity: 4,
        interval: 50,
        timeout: 100,
        over: function(){
          $(this).trigger("open");
        },
        out: function(){
          $(this).trigger("close");
        }
      };

      //for each top level menu item tab, attach a hoverIntent event and bind another event
      topLevel.each(function(){
        var topLevelTab = $(this);
        var tabType = topLevelTab.attr('data-menu');
        var currentSubnav = subnav.find('.' + tabType + '.second_level').parent();
        var enter = function(){topLevelTab.trigger("mouseenter");};
        var leave = function(){setTimeout(topLevelTab.trigger("mouseout"), 200)};
        currentSubnav.hover(function(){enter()},function(){leave()});
        topLevelTab.bind("open", function(){
            //make the tab "highlight" on hover
            $(this).addClass('selected');
            activeSubnav.hide();
            //show the current subnav and trigger it to stay open
            currentSubnav.slideDown('fast');
            topLevelTab.trigger("hovering");
          }).bind("hovering", function(){
            currentSubnav.show();
          })
          .bind("close", function(){
            //take away tab highlight
            $(this).removeClass('selected');
            activeTab.addClass('selected');
            activeSubnav.show();
            if(!$(this).hasClass('active')) {
              //the stuff to do to if it's not the current tab
              currentSubnav.slideUp('fast');
            }
          }).hoverIntent(hoverSettings);

      });
    },
    thirdLevelNavSetup : function(){
        var children = $('.third_level:first-child');

        $.each(children, function(i, item) {
            var child = $(item);
            var li = child.parent().parent();
            var  ul = child.parent();

            li.prepend($('<div class="arrow_icon_menu"></div>'));
            li.hover(
                function(){
                    ul.slideDown('fast');
                },
                function(){
                    ul.slideUp('fast');
            });
        });

    }
  };
})();

File: app/lib/experimental/navigation/menu.rb
#
# Copyright 2013 Red Hat, Inc.
#
# This software is licensed to you under the GNU General Public
# License as published by the Free Software Foundation; either version
# 2 of the License (GPLv2) or (at your option) any later version.
# There is NO WARRANTY for this software, express or implied,
# including the implied warranties of MERCHANTABILITY,
# NON-INFRINGEMENT, or FITNESS FOR A PARTICULAR PURPOSE. You should
# have received a copy of GPLv2 along with this software; if not, see
# http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.


module Experimental
  module Navigation
    class Menu

      attr_accessor :key, :display, :type, :items

      # Initalizer for the Navigation Menu object
      def initialize(*args)
        process_additions(*args)
        generate
      end

      # Returns whether this item is accessible based on authorization rules
      #   Expects either a Proc or a boolean
      def accessible?
        if @authorization.is_a? Proc
          @authorization.call
        else
          @authorization
        end
      end

      # Dynamically sets the authorization rule
      def authorization=(authorization)
        @authorization = authorization
      end

      # Defines the JSON structure for navigation menus
      #
      # @return [String] the JSON representation of a navigation menu
      def as_json(*args)
        item = {
          :key    => @key,
          :display=> @display,
          :type   => @type,
          :items  => @items
        }

        return item
      end

      # Generates the menu structure
      def generate
        prune
      end

      def process_additions(*args)
        additions = Experimental::Navigation::Additions.list
        additions.each do |addition|

          index =  @items.index{|item| item.key.to_sym == addition[:key].to_sym}
          if index && addition[:type] == :delete
            @items.delete_at(index)
          elsif index
            index += 1 if (addition[:type] == :after)
            node = addition[:node].new(*args)
            @items.insert(index, node)
          end
        end

      end

      # Recursively prunes the menu items by checking if they are accessible
      def prune
        @items.delete_if do |item|
          if item.accessible?
            if item.is_a? Experimental::Navigation::Menu
              item.prune
              item.items.empty?
            else
              false
            end
          else
            true
          end
        end
      end

    end
  end
end
